include/ginkgo/core/base/range_accessor_references.hpp

```diff
@@ -0,0 +1,429 @@
+        arithmetic_type                                                    \
+        _oper(Reference &&ref1, const Reference &ref2)                     \
+    {                                                                      \
+        return std::move(ref1) = to_value_type<arithmetic_type>(ref1)      \
+                   _op to_value_type<arithmetic_type>(ref2);               \
+    }                                                                      \
+    friend GKO_ENABLE_REFERENCE_CONSTEXPR GKO_INLINE GKO_ATTRIBUTES        \
+        arithmetic_type                                                    \
+        _oper(Reference &&ref, const arithmetic_type &a)                   \
+    {                                                                      \
+        return std::move(ref) = to_value_type<arithmetic_type>(ref) _op a; \
+    }
+
+    GKO_REFERENCE_ASSIGNMENT_OPERATOR_OVERLOAD(operator*=, *)
+    GKO_REFERENCE_ASSIGNMENT_OPERATOR_OVERLOAD(operator/=, /)
+    GKO_REFERENCE_ASSIGNMENT_OPERATOR_OVERLOAD(operator+=, +)
+    GKO_REFERENCE_ASSIGNMENT_OPERATOR_OVERLOAD(operator-=, -)
+#undef GKO_REFERENCE_ASSIGNMENT_OPERATOR_OVERLOAD
+
+    friend GKO_ENABLE_REFERENCE_CONSTEXPR GKO_INLINE GKO_ATTRIBUTES
+        arithmetic_type
+        operator-(const Reference &ref)
+    {
+        return -to_value_type<arithmetic_type>(ref);
+    }
+};
+
+// There is no more need for this macro in this file
+#undef GKO_ENABLE_REFERENCE_CONSTEXPR
+
+
+}  // namespace detail
+
+
+/**
+ * This namespace contains reference classes used inside accessors.
+ *
+ * @warning These classes should only be used by accessors.
+ */
+namespace reference_class {
+
+
+/**
+ * Reference class for a different storage than arithmetic type. The
+ * conversion between both formats is done with a simple static_cast.
+ *
+ * Copying this reference is disabled, but move construction is possible to
+ * allow for an additional layer (like gko::range).
+ * The assignment operator only works for an rvalue reference (&&) to
+ * prevent accidentally copying the reference and working on a reference.
+ *
+ * @tparam ArithmeticType  Type used for arithmetic operations, therefore,
+ *                         the type which is used for input and output of this
+ *                         class.
+ *
+ * @tparam StorageType  Type actually used as a storage, which is converted
+ *                      to ArithmeticType before usage
+ */
+template <typename ArithmeticType, typename StorageType>
+class reduced_storage
+    : public detail::enable_reference_operators<
+          reduced_storage<ArithmeticType, StorageType>, ArithmeticType> {
+public:
+    using arithmetic_type = std::remove_cv_t<ArithmeticType>;
+    using storage_type = StorageType;
+
+    // Allow move construction, so perfect forwarding is possible (required
+    // for `range` support)
+    reduced_storage(reduced_storage &&) = default;
+
+    reduced_storage() = delete;
```

Comment:
not: I think you don't need to delete the default constructor if there is a constructor with parameters

Reply:
I think I need it. According to [cppreference/default_constructor](https://en.cppreference.com/w/cpp/language/default_constructor):
>If no user-defined constructors are present and the implicitly-declared default constructor is not trivial, the user may still inhibit the automatic generation of an implicitly-defined default constructor by the compiler with the keyword delete. 

I think the implicitly-declared constructor is probably trivial (assigning a `nullptr` to a `storage_type *const`). In any case, I like to specify it here since the rules are a bit complicated, so explicitly stating that I don't want to have a default constructor felt reasonable, especially since a reference to nothing does not make sense.